doc: Update mpirun map/rank/bind to reflect PRRTE (#11026)
open-mpi/ompi: docs/man-openmpi/man1/mpirun.1.rst

* ``--map-by <object>``: Map to the specified object, defaults to
  socket. Supported options include ``slot``, ``hwthread``, ``core``,
  ``L1cache``, ``L2cache``, ``L3cache``, ``package``, ``numa``,
  ``node``, ``seq``, ``rankfile``, ``pe-list=#``, and ``ppr``. 

>   Note that ``socket`` is also accepted as an alias for ``package``.

Author: Good suggestion. Will add this in the next push.

  Any object can include modifiers by adding a ``:`` and any combination

>   of ``pe=n`` (bind n processing elements to each proc), ``span``

Author: Will do

* ``--map-by <object>``: Map to the specified object, defaults to
  socket. Supported options include ``slot``, ``hwthread``, ``core``,
  ``L1cache``, ``L2cache``, ``L3cache``, ``package``, ``numa``,
  ``node``, ``seq``, ``rankfile``, ``pe-list=#``, and ``ppr``. 
  Any object can include modifiers by adding a ``:`` and any combination
  of the following:

>     * ``pe=n`` : bind n processing elements to each proc
>     * ``span`` : load balance the processes across the allocation
>     * ``oversubscribe`` : allow more processes on a node than processing elements
>     * ``nooversubscribe`` : (default)
>     * ``nolocal`` : do not place processes on the same host as the `mpirun` process
>     * ``hwtcpus`` : use hardware threads as CPU slots for mapping
>     * ``corecpus`` : use processor cores as CPU slots for mapping (Default)
>     * ``file=`` : used with ``rankfile`` to specify the file to use
>     * ``ordered`` : used with ``pe-list`` to bind each process to one of the specifed processing elements

Reviewer: Nit picky suggestions:

```suggestion
    * ``pe=n``: bind ``n`` processing elements to each proc
    * ``span``: load balance the processes across the allocation
    * ``oversubscribe``: allow more processes on a node than processing elements
    * ``nooversubscribe``: do *not* allow more processes on a node than processing elements (default)
    * ``nolocal``: do not place processes on the same host as the ``mpirun`` process
    * ``hwtcpus``: use hardware threads as CPU slots for mapping
    * ``corecpus``: use processor cores as CPU slots for mapping (default)
    * ``file=filename``: used with ``rankfile``; use ``filename`` to specify the file to use
    * ``ordered``: used with ``pe-list`` to bind each process to one of the specified processing elements
```

* If ``--map-by OBJ1 --bind-to OBJ2`` then
    * ``--map-by`` specified ``OBJ1``
    * ``--bind-to`` specified ``OBJ2``
    * ``--rank-by`` defaults to ``fill``
    * Result: ``--map-by OBJ2 --bind-to OBJ2 --rank-by fill``

Consider 2 identical hosts (``hostA`` and ``hostB``) with 2 sockets (denoted by ``[]``) each with 8 cores (denoted by ``/../``) and two hardware threads per core (denoted by a ``.``).

Default of ``--map-by core --bind-to core --rank-by span`` when the number of processes is less than or equal to 2.

.. code::

  shell$ mpirun --np 2 --host hostA:4,hostB:2 ./a.out
  R0  hostA  [BB/../../../../../../..][../../../../../../../..]
  R1  hostA  [../BB/../../../../../..][../../../../../../../..]

Default of ``--map-by package --bind-to package --rank-by fill`` when the number of processes is greater than 2.

.. code::

  shell$ mpirun --np 4 --host hostA:4,hostB:2 ./a.out
  R0  hostA  [BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../..]
  R1  hostA  [BB/BB/BB/BB/BB/BB/BB/BB][../../../../../../../..]
  R2  hostA  [../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB]
  R3  hostA  [../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/BB]

> If only ``--map-by OBJ`` is specified then it implies ``--bind-to OBJ --rank-by fill``. The example below results in ``--map-by hwthread --bind-to hwthread --rank-by fill``

Reviewer: ```suggestion
If only ``--map-by OBJ`` is specified, then it implies ``--bind-to OBJ --rank-by fill``. The example below results in ``--map-by hwthread --bind-to hwthread --rank-by fill``
```

.. code::

  shell$ mpirun --np 4 --map-by hwthread --host hostA:4,hostB:2 ./a.out
  R0  hostA  [B./../../../../../../..][../../../../../../../..]
  R1  hostA  [.B/../../../../../../..][../../../../../../../..]
  R0  hostA  [../B./../../../../../..][../../../../../../../..]
  R1  hostA  [../.B/../../../../../..][../../../../../../../..]

> If only ``--bind-to OBJ`` is specified then ``--map-by`` is determined by the number of processes and ``--rank-by fill``. The example below results in ``--map-by package --bind-to core --rank-by fill``

Reviewer: ```suggestion
If only ``--bind-to OBJ`` is specified, then ``--map-by`` is determined by the number of processes and ``--rank-by fill``. The example below results in ``--map-by package --bind-to core --rank-by fill``
```

* If no options are specified then

    * If the number of processes is less than or equal to 2, then:

        * ``--map-by`` is ``core``
        * ``--bind-to`` is ``core``
        * ``--rank-by`` is ``span``
        * Result: ``--map-by core --bind-to core --rank-by span``

    * Otherwise:

        * ``--map-by`` is ``package``
        * ``--bind-to`` is ``package``
        * ``--rank-by`` is ``fill``
        * Result: ``--map-by package --bind-to package --rank-by fill``

* If only ``--map-by OBJ`` (where ``OBJ`` is something like ``core``) is specified, then:

    * ``--map-by`` specified ``OBJ``
    * ``--bind-to`` uses the same ``OBJ`` as ``--map-by``
    * ``--rank-by`` defaults to ``fill``
    * Result: ``--map-by OBJ --bind-to OBJ --rank-by fill``

* If only ``--bind-to OBJ`` (where ``OBJ`` is something like ``core``) is specified, then:

    * ``--map-by`` is either ``core`` or ``package`` depending on the number of processes
    * ``--bind-to`` specified ``OBJ``
    * ``--rank-by`` defaults to ``fill``
    * Result: ``--map-by OBJ --bind-to OBJ --rank-by fill``

> * If ``--map-by OBJ1 --bind-to OBJ2``, then:

Reviewer: RST requires a blank line after this line (before the sub-bullets).

^^^^^^^^^^^^^^^^^^^

Processes in the MPI application inherit their environment from the
PRRTE daemon upon the node on which they are running.  The
environment is typically inherited from the user's shell.  On remote
nodes, the exact environment is determined by the boot MCA module
used.  The rsh launch module, for example, uses either rsh/ssh to

> launch the PRTE daemon on remote nodes, and typically executes one

Reviewer: ```suggestion
launch the PRRTE daemon on remote nodes, and typically executes one
```

> or more of the user's shell-setup files before launching the PRTE

Reviewer: ```suggestion
or more of the user's shell-setup files before launching the PRRTE
```
